activerecord/lib/active_record/relation/calculations.rb

```diff
@@ -389,6 +394,11 @@ def execute_grouped_calculation(operation, column_name, distinct) #:nodoc:
           result[key] = type_cast_calculated_value(row[column_alias], operation) do |value|
             type ||= column.try(:type_caster) ||
               lookup_cast_type_from_join_dependencies(column_name.to_s) || Type.default_value
+
+            if operation == "average" && type.is_a?(Type::Integer)
+              type = Type::Decimal.new
```

Comment:
ditto https://github.com/rails/rails/pull/41418#discussion_r575094917

Reply:
`type` comes in from above. It will be kept while looping. Isn't that an issue?

Reply:
It is intentionally kept since the block will be evaluated multiple times and `lookup_cast_type_from_join_dependencies` is a bit expensive.

Reply:
Ok. I took in the second (identical) proposal as well because memoization is only there for expensive `lookup_cast_type_from_join_dependencies`.